### PR TITLE
Always check if shard is local when accessing its stats

### PR DESCRIPTION
**Pull request by otoolep**

### CODE REVIEW

**Comment by otoolep** [comment]
Fixes issue https://github.com/influxdb/influxdb/issues/2452

File: server.go [hunk 1]
```diff
@@ -3355,7 +3359,7 @@ func (s *Server) DiagnosticsAsRows() []*influxql.Row {
 			nodes = append(nodes, strconv.FormatUint(n, 10))
 		}
 		var path string
-		if sh.store != nil {
```
Comment (otoolep) [inline server.go]: Not strictly necessary, but standardizes on using `HasDataNodeID` to check if a shard is local.

File: cmd/influxd/server_integration_test.go [hunk 2]
```diff
@@ -1461,7 +1461,7 @@ func TestSingleServerDiags(t *testing.T) {
 	config := main.NewConfig()
 	config.Monitoring.Enabled = true
 	config.Monitoring.WriteInterval = main.Duration(100 * time.Millisecond)
-	nodes := createCombinedNodeCluster(t, testName, dir, 1, config)
+	nodes := createCombinedNodeCluster(t, testName, dir, 3, config)
```
Comment (otoolep) [inline cmd/influxd/server_integration_test.go]: This change ensures diags and stats self-monitoring works for both single and multiple node cluster.

File: shard.go [hunk 3]
```diff
@@ -234,7 +234,9 @@ func (s *Shard) close() error {
 	if s.store != nil {
 		_ = s.store.Close()
 	}
-	s.stats.Inc("close")
```
Comment (otoolep) [inline shard.go]: Also caused panic. We need to continue to rework the shard type in our code. @corylanou started some of this work.

Comment (corylanou) [inline shard.go]: I'm not sure how stats can ever be nil.  We instantiate it for both creating or opening a shard.  It feels like somewhere else in our code we are creating a shard and not instantiating it properly?

Comment (otoolep) [inline shard.go]: Yes, this is why we need to continue to rework this code. Here:

https://github.com/influxdb/influxdb/blob/master/server.go#L318

we add shards, but never open them. In this function nothing prevents `close` being called on a shard object that was never opened.

Comment (otoolep) [inline shard.go]: https://github.com/influxdb/influxdb/blob/master/server.go#L275 is where shards can be closed that were never opened.

Comment (corylanou) [inline shard.go]: The other problem it would appear is something is calling code when the server is shut down, which doesn't seem right either.

**Comment by corylanou** [comment]
+1

**Comment by otoolep** [comment]
Thanks @corylanou -- I have opened ticket https://github.com/influxdb/influxdb/issues/2455 to work on refactoring.

**Pull request merged**